File: combination-sum.java
class Solution {    
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> combinations = new ArrayList<>();
        combinationSum(candidates, 0, target, combinations, new ArrayList<>());
        return combinations;
    }
    
    private void combinationSum(int[] candidates, int i, int target, List<List<Integer>> combinations, List<Integer> currCombination) {
        if (target == 0) {
            combinations.add(currCombination);
            return;
        }
        
        if (i >= candidates.length || target < 0) {
            return;
        }
        
        List<Integer> copy = new ArrayList<>(currCombination);
        copy.add(candidates[i]);
        combinationSum(candidates, i, target - candidates[i], combinations, copy);
        combinationSum(candidates, i + 1, target, combinations, currCombination);
    }
}

File: combination-sum/combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        
        ans = []
        def backtrack(i, curr, seq):
            if curr > target:
                return
            
            if curr == target:
                ans.append(seq[:])
                return
            
            if i == len(candidates):
                return
            
            for j in range(i, len(candidates)):
                seq.append(candidates[j])
                backtrack(j, curr + candidates[j], seq)
                seq.pop()
        
        backtrack(0, 0, [])
        
        return ans